### PR TITLE
feat: build safenode with custom features

### PR DESCRIPTION
**Pull request by RolandSherwin**
Ontop of #52

### CODE REVIEW

File: src/ansible.rs [hunk 1]
```diff
@@ -140,9 +140,6 @@ impl AnsibleRunnerInterface for AnsibleRunner {
             .map(|(name, vars)| (name, vars.ansible_host))
             .collect();
 
-        if inventory.is_empty() {
```
Comment (jacderida) [inline src/ansible.rs:143]: Why would you not want an error for empty inventory?

File: src/deploy.rs [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+            testnet_deploy,
+            name,
+            node_count,
+            vm_count,
+            logstash_details,
+            sn_codebase_type,
+        }
+    }
+
+    pub async fn deploy(self) -> Result<()> {
```
Comment (jacderida) [inline src/deploy.rs:42]: Minor style suggestion: if the class is a command, the term `execute` is often associated with running the command.

https://en.wikipedia.org/wiki/Command_pattern

The following:
```
let cmd = DeployCmd { };
cmd.execute();
```
Reads a little clearer than:
```
let cmd = DeployCmd { };
cmd.deploy();
```

File: src/lib.rs [hunk 3]
```diff
@@ -42,10 +43,29 @@ use std::{
...
 use tar::Archive;
 
+/// How or where to build the binaries from.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum SnCodebaseType {
+    Main {
+        // CSV of features to enable on main::safenode
+        safenode_features: Option<String>,
+    },
+    CustomBranch {
```
Comment (jacderida) [inline src/lib.rs:57]: Minor style suggestion: I think we could drop the `Custom` part now.

File: src/lib.rs [hunk 4]
```diff
@@ -42,10 +43,29 @@ use std::{
...
+        repo_owner: String,
+        branch: String,
+        // CSV of features to enable on the custom branch::safenode
+        safenode_features: Option<String>,
+    },
+    PreBuiltBinary {
```
Comment (jacderida) [inline src/lib.rs:63]: Minor style suggestion: I think `Versioned` would be a bit more specific.

File: src/lib.rs [hunk 5]
```diff
@@ -42,10 +43,29 @@ use std::{
...
     process::{Command, Stdio},
-    time::{Duration, Instant},
+    time::Duration,
 };
 use tar::Archive;
 
+/// How or where to build the binaries from.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub enum SnCodebaseType {
+    Main {
```
Comment (jacderida) [inline src/lib.rs:53]: It might be worth putting some doc comments on this entry just to say that if features are used, the binary will be built, but otherwise the 'latest' will be fetched from S3.

File: src/main.rs [hunk 6]
```diff
@@ -47,30 +48,40 @@ enum Commands {
     },
     /// Deploy a new testnet environment using the latest version of the safenode binary.
     Deploy {
-        /// Optionally supply the name of a branch on the Github repository to be used for the
```
Comment (jacderida) [inline src/main.rs:50]: Minor suggestion: I think it's useful to have the arguments sorted in alphabetical order rather than arbitrarily.

File: src/main.rs [hunk 7]
```diff
@@ -47,30 +48,40 @@ enum Commands {
...
+        vm_count: u16,
         /// The cloud provider to deploy to.
         ///
         /// Valid values are "aws" or "digital-ocean".
         #[clap(long, default_value_t = CloudProvider::DigitalOcean, value_parser = parse_provider, verbatim_doc_comment)]
         provider: CloudProvider,
+        /// The name of the Logstash stack to forward logs to.
+        #[clap(long, default_value = "main")]
+        logstash_stack_name: String,
+        /// The features to enable on the safenode instance
```
Comment (jacderida) [inline src/main.rs:70]: Minor suggestion: use "binary" here rather than "instance".

File: src/ansible.rs (hunk 1 above)
Comment (RolandSherwin) [reply src/ansible.rs:143]: During `list_inventory`, we don't have info about whether the `build` machine was created/not during `deploy`. 
Thus, it is better to check for `empty invetory` after the `inventory_list() fn` is called instead of propagating error from it. 

The other option is to handle the error case.
Or we should not be writing `.{name}_build_inventory_digital_ocean.yml` file if the build vm is not setup. I haven't tried going that route as it was not obvious to me.

File: src/deploy.rs (hunk 2 above)
Comment (RolandSherwin) [reply src/deploy.rs:42]: That is indeed cleaner! Thanks.

File: src/ansible.rs (hunk 1 above)
Comment (jacderida) [reply src/ansible.rs:143]: OK no problem.

File: src/main.rs (hunk 6 above)
Comment (RolandSherwin) [reply src/main.rs:50]: I organized it based on the `required, optional args` + `bundled together related args`. It is a bit arbitrary, but I was able to follow it better as the related ones were close by. 

I can change it, if it's hard to follow.

Comment (jacderida) [reply src/main.rs:50]: Nah it's ok, not a problem.

**Review by jacderida — APPROVED** [review]
